composition-js/src/composeDirectiveManager.ts

```diff
@@ -0,0 +1,366 @@
+import { assert, CoreFeature, DirectiveDefinition, Subgraphs, ERRORS, SubgraphASTNode, didYouMean, suggestionList } from '@apollo/federation-internals';
+import { ASTNode, GraphQLError } from 'graphql';
+import { CompositionHint, HINTS } from './hints';
+
+const setItemOnObject = <T>(key: string, t: T, obj: { [key: string]: T[]}) => {
+  if (key in obj) {
+    obj[key].push(t);
+  } else {
+    obj[key] = [t];
+  }
+};
+
+const originalDirectiveName = (directive: DirectiveDefinition, feature: CoreFeature): string => {
```

Comment:
If I put it into CoreFeature, what should the return value be? Does it make sense to always return the directive name or should I check to make sure that the definition actually belongs to this feature?

Reply:
> If I put it into CoreFeature, what should the return value be?

My intent is that the method would check if the provided directive belongs to the feature like `CoreFeature.sourceFeature` does, so check if the name is either prefixed with the feature name, the exact feature name, or is part of the import (as a "as"), and if doesn't match those, would return `undefined`.

> Does it make sense to always return the directive name or should I check to make sure that the definition actually belongs to this feature?

Do you mean, what if the schema defines a directive called `@federation__foobar`, despite the `federation` spec clearly not having a `foobar` directive?

If so, then I think it's fine that such method don't worry about this (and would thus just say that `foobar` is the orignal name), the same way that the existing `CoreFeatures.sourceFeature` would return the `federation` feature on such directive at the moment.

Which is not me saying that we should never validate that a schema don't tries to lie about a directive belonging to a feature, just that such validation belongs earlier in the process and the methods of `CoreFeature(s)` can assume it's already been validated. In practice, we do validate this for the `federation` spec early in the process, but it's true we don't do it for "custom" core feature because well, we cannot really at this stage since we don't look-up feature definitions in the first place. We'll need to improve that, but it's work for another day :).